### DISCUSSION TITLE
Support for Python's `~=` version specifier

### DISCUSSION THREAD

**Discussion by slmtpz** [renovatebot/renovate#8280] [Request Help] [answered]
Hi! We recently started POC'ing Renovate for our Python packages. However, we couldn't tackle the issue where Renovate doesn't respect [`~=` version specifier](https://www.python.org/dev/peps/pep-0440/#compatible-release) in Python. We tried several different configurations. We might have not tried the correct configuration and that's why I wanted to discuss this with the experts.

The way we specify the versions in `setup.py` is like following:

```
"package-one~=1.2",
"package-two~=0.2.3",
```
In more elaborate terms, if a package has a stable version (major > 0), then pin its major version and require equal or bigger minor version. If the major version of the specified version is 0, then pin its both major and minor version and require equal or bigger patch version.

The versions that I want Renovate to suggest to us are like the following:

```
"package-one~=1.2", -> `~=1.5`
"package-two~=0.2.3", -> `~=0.2.6`
```

whilst the actual behaviour is as follows:

```
"package-one~=1.2", -> `~=2.0.4`
"package-two~=0.2.3", -> `~=1.0.2`
```

The behaviour is the same with and without `"rangeStrategy": "bump"` configuration.

Thus, my two questions are:
- How do I make Renovate conform to `~=` version specifier and obey the semver?
- Why does it suggest a version with all 3 (major, minor, patch) versions even though the original version specifier was `~=1.2`?

Thanks in advance.

**Comment by rarkins** [comment, marked as answer]
Renovate already has support for Python's `~=` operator, although there's always a possibility there's a bug. You can see existing tests in this test file: https://github.com/renovatebot/renovate/blob/master/lib/versioning/pep440/index.spec.ts

For any language, if you want Renovate to update a constraint when the new version *satisfies* the existing range, then you must configure `"rangeStrategy": "bump"`. You'll never get an in-range update with the default rangeStrategy=replace.

If you want *patch* updates when minor updates also exist then you also need to configure `"separateMinorPatch": true`. Otherwise default behaviour is to propose the latest non-major update.

Can you configure both of these and create a [minimal reproduction repo](https://github.com/renovatebot/renovate/blob/master/docs/development/minimal-reproductions.md) to demonstrate if it's a bug?
